lib/src/package_config_impl.dart

```diff
@@ -0,0 +1,178 @@
+        // _SimplePackage validates these properties.
+        try {
+          _validatePackageData(package.name, package.root,
+              package.packageUriRoot, package.languageVersion);
+        } catch (e) {
+          throw ArgumentError.value(
+              packages, "packages", "Package ${package.name}: ${e.message}");
+        }
+      }
+      var name = package.name;
+      if (result.containsKey(name)) {
+        throw ArgumentError.value(name, "packages", "Duplicate package name");
+      }
+      result[name] = package;
+    }
+
+    // Check that no root URI is a prefix of another.
+    if (result.length > 1) {
+      // Uris cache their toString, so this is not as bad as it looks.
+      var rootUris = [...result.values]
+        ..sort((a, b) => a.root.toString().compareTo(b.root.toString()));
+      var prev = rootUris[0];
+      var prevRoot = prev.root.toString();
+      for (int i = 1; i < rootUris.length; i++) {
+        var next = rootUris[i];
+        var nextRoot = next.root.toString();
+        // If one string is a prefix of another,
+        // the former sorts just before the latter.
+        if (nextRoot.startsWith(prevRoot)) {
+          throw ArgumentError.value(
+              packages,
+              "packages",
+              "Package ${next.name} root overlaps "
+                  "package ${prev.name} root.\n"
+                  "${prev.name} root: $prevRoot\n"
+                  "${next.name} root: $nextRoot\n");
+        }
+        prev = next;
+      }
+    }
+    return result;
+  }
+
+  Iterable<Package> get packages => _packages.values;
+
+  Package /*?*/ operator [](String packageName) => _packages[packageName];
+
+  /// Provides the associated package for a specific [file] (or directory).
+  ///
+  /// Returns a [Package] which contains the [file]'s path.
+  /// That is, the [Package.rootUri] directory is a parent directory
+  /// of the [file]'s location.
+  /// Returns `null` if the file does not belong to any package.
+  Package /*?*/ packageOf(Uri file) {
+    String path = file.toString();
+    for (var package in _packages.values) {
+      var rootPath = package.root.toString();
+      if (path.startsWith(rootPath)) return package;
+    }
+    return null;
+  }
+
+  Uri /*?*/ resolve(Uri packageUri) {
+    String packageName = checkValidPackageUri(packageUri);
+    return _packages[packageName]?.packageUriRoot?.resolveUri(
+        Uri(path: packageUri.path.substring(packageName.length + 1)));
+  }
+
+  Uri /*?*/ toPackageUri(Uri nonPackageUri) {
+    if (nonPackageUri.isScheme("package")) {
+      throw ArgumentError.value(
+          nonPackageUri, "nonPackageUri", "Must not be a package URI");
+    }
+    if (nonPackageUri.hasQuery || nonPackageUri.hasFragment) {
+      throw ArgumentError.value(nonPackageUri, "nonPackageUri",
+          "Must not have query or fragment part");
+    }
+    for (var package in _packages.values) {
+      var root = package.root;
```

Comment:
Shouldn't this be `package.packageUriRoot`?

Reply:
It probably should, yes.